[Management] Allows for imports to select existing index

chrisronline opened the pull request:
Replaces #13787
Relates to #13359

Builds upon @tylersmalley's functional changes, but makes some changes to the UI so there is a single modal that shows a table of conflicted index pattern ids, grouped by the id and provides a way to apply a change to an entire set of affected saved objects.

<img width="845" alt="screen shot 2017-09-22 at 4 33 06 pm" src="https://user-images.githubusercontent.com/56682/30763362-d2dca604-9fb3-11e7-8991-a120cd1b4cd0.png">

/cc @alexfrancoeur

tylersmalley commented:
This is a huge win @chrisronline, thanks for taking it over! I should be able to review early next week.

cjcenizal commented on src/core_plugins/kibana/public/management/sections/objects/change_index_modal.js:
```diff
@@ -0,0 +1,207 @@
+  };
+
+  onKeyDown = (event) => {
+    if (event.keyCode === 27) { // ESC key
```
You can remove the comment here and instead import `keyCodes` from the UI Framework:

```javascript
import { keyCodes } from 'ui_framework/services';

if (event.keyCode === keyCodes.ESCAPE) {
```

cjcenizal commented on src/core_plugins/kibana/public/management/sections/objects/change_index_modal.js:
```diff
@@ -0,0 +1,207 @@
+    event.persist();
+    this.setState(state => {
+      return {
+        ...state,
```
Why spread state here?

chrisronline replied:
Good point. Unnecessary

cjcenizal commented on src/core_plugins/kibana/public/management/sections/objects/change_index_modal.js:
```diff
@@ -0,0 +1,207 @@
+                    Sample of Affected Objects
+                  </KuiTableHeaderCell>
+                  <KuiTableHeaderCell width="200">
+                    New Index
```
Do we want this to say "New index pattern" instead? (Note the sentence casing too).

epixa commented:
All the tests passed.  The build failure was just the result of s3 upload failure for the success screenshots.

cjcenizal commented on src/core_plugins/kibana/public/management/sections/objects/change_index_modal.js:
```diff
@@ -0,0 +1,207 @@
+              data-test-subj="cloneConfirmButton"
+              onClick={this.changeIndex}
+            >
+              Confirm All Changes
```
Can we sentence case this? "Confirm all changes". I think this is how we should case our button text but I'm double-checking with our copywriters.

cjcenizal commented on src/core_plugins/kibana/public/management/sections/objects/change_index_modal.js:
```diff
@@ -0,0 +1,207 @@
+                    Count
+                  </KuiTableHeaderCell>
+                  <KuiTableHeaderCell>
+                    Sample of Affected Objects
```
Can we change this to sentence case: "Sample of affected objects"

cjcenizal commented on src/core_plugins/kibana/public/management/sections/objects/change_index_modal.js:
```diff
@@ -0,0 +1,207 @@
+          </KuiModalHeader>
+          <KuiModalBody>
+            <KuiModalBodyText>
+              <p>The following saved objects use index patterns that do not exist.</p>
```
Can we add another sentence here explaining what the user is supposed to do?

```html
<p>
  The following saved objects use index patterns that do not exist.
  Please select the index patterns you'd like to re-associate them with.
</p>
```

chrisronline commented:
Thanks for the comments and suggestions @cjcenizal! I've updated the branch with the changes.

alexfrancoeur commented:
@chrisronline I just attempted to import an older set of dashboards and visualizations where I know I do not have the index pattern defined. I was allowed to import the index pattern but was not prompted with a modal window to replace the index pattern. In my environment I have other index patterns defined and visualizations already there, so it was not an empty environment. I'm not sure if that would affect the behavior at all. The end result ends up being a dashboard with errors associated with the visualizations and no easy way to bulk modify them. 

<img width="1309" alt="screen shot 2017-09-25 at 3 13 19 pm" src="https://user-images.githubusercontent.com/25181823/30826299-1cfa6d3c-a204-11e7-8244-2f3abefabd0a.png">

The error no longer provides the option to edit or create a new index pattern.

<img width="1307" alt="screen shot 2017-09-25 at 3 16 35 pm" src="https://user-images.githubusercontent.com/25181823/30826434-9827cae0-a204-11e7-8aa5-1c599c4aec22.png">

Oddly enough, I have no problem when I attempt to import a single saved search.

<img width="950" alt="screen shot 2017-09-25 at 3 22 47 pm" src="https://user-images.githubusercontent.com/25181823/30826698-7bc622d8-a205-11e7-811c-557caf8715ab.png">

Zip file with older exported saved objects [here](https://github.com/elastic/kibana/files/1330609/Metricbeat.TSVB.2.zip).

Also, have we thought about how we want to handle if the user has not already created an index pattern? Should we offer a "create index pattern with this ID" as an option from the drop down that would lead them to the index pattern management wizard with a pre-populated ID?

chrisronline commented:
Thanks @alexfrancoeur! I'm not sure, but I will use that dataset and investigate!

alexfrancoeur commented:
After conversations in Slack, we are not prompted with the modal because these are TSVB visualizations which do not use an index pattern id. We can ignore my error. Sorry about that @chrisronline! Thoughts on my second comment?

> Also, have we thought about how we want to handle if the user has not already created an index pattern? Should we offer a "create index pattern with this ID" as an option from the drop down that would lead them to the index pattern management wizard with a pre-populated ID?

chrisronline commented:
<img width="785" alt="screen shot 2017-09-25 at 5 01 13 pm" src="https://user-images.githubusercontent.com/56682/30830775-29f9c19a-a213-11e7-9e70-91e1eef4f74e.png">

@alexfrancoeur What if we presented this option in the dropdown and they can handle it afterwards? I think we want a more complete solution for this suggestion, like hooking up the same flow on the visualization and dashboard pages to indicate there is a missing index pattern id and provide a flow to create one. Maybe providing a simple link in the modal might help too?

tylersmalley commented on src/core_plugins/kibana/public/management/sections/objects/_objects.js:
```diff
@@ -178,71 +193,85 @@ uiModules.get('apps/management')
         }
 
-        return new Promise((resolve) => {
-          confirmModal(
```
It looks like the functionality to choose if you want to be prompted to automatically overwrite, or be prompted each time was removed.

chrisronline replied:
Hm. I thought we wanted that behavior removed? Is that not the case anymore?

tylersmalley commented on src/core_plugins/kibana/public/management/sections/objects/_objects.js:
```diff
@@ -178,71 +193,85 @@ uiModules.get('apps/management')
+              obj.id = doc._id;
+              return obj.applyESResp(doc)
+                .then(() => {
+                  return obj.save({ confirmOverwrite : false });
```
Here we're preventing the ability to overwrite the objects, is that the desired behavior?

chrisronline replied:
Because I thought we wanted to remove that modal, I changed all the behavior to default to not overrwrite all, but it seems this is wrong regardless.

tylersmalley commented on src/core_plugins/kibana/public/management/sections/objects/change_index_modal.js:
```diff
@@ -0,0 +1,211 @@
+            </KuiModalBodyText>
+            <KuiControlledTable>
+              <KuiToolBar>
+                <KuiToolBarSection>
```
Can we eliminate this row unless there are multiple pages?

chrisronline replied:
I tried it a few different ways and not sure what's the best visuals:

This is now:
<img width="801" alt="screen shot 2017-09-26 at 1 18 00 pm" src="https://user-images.githubusercontent.com/56682/30874087-5660f67c-a2bd-11e7-9daf-2ff26ee9c004.png">

This is if I don't show the toolbar:
<img width="802" alt="screen shot 2017-09-26 at 1 19 23 pm" src="https://user-images.githubusercontent.com/56682/30874097-5deb11b6-a2bd-11e7-94e6-bfcdb0ac1668.png">

The best way is probably to only do the controlled table but that seems like a weird thing to read later?

cjcenizal replied:
It should definitely be possible to render a table without the toolbar at the top. You'll just have to take it out of the controlled table component. There should be an example in the doc site, I think?

tylersmalley commented:
I am also getting an error on import when there are no existing index patterns:

```
TypeError: Cannot read property 'id' of undefined
    at http://localhost:5601/bundles/kibana.bundle.js?v=8467:332390:47
    at http://localhost:5601/bundles/commons.bundle.js?v=8467:3786:25
    at http://localhost:5601/bundles/commons.bundle.js?v=8467:3464:16
    at baseForOwn (http://localhost:5601/bundles/commons.bundle.js?v=8467:2437:15)
    at http://localhost:5601/bundles/commons.bundle.js?v=8467:3785:10
    at new ChangeIndexModal (http://localhost:5601/bundles/kibana.bundle.js?v=8467:332388:39)
    at http://localhost:5601/bundles/kibana.bundle.js?v=8467:128129:19
    at measureLifeCyclePerf (http://localhost:5601/bundles/kibana.bundle.js?v=8467:127910:13)
    at ReactCompositeComponentWrapper._constructComponentWithoutOwner (http://localhost:5601/bundles/kibana.bundle.js?v=8467:128128:17)
    at ReactCompositeComponentWrapper._constructComponent (http://localhost:5601/bundles/kibana.bundle.js?v=8467:128114:22)
```

Adding to the discussion above, how about having an option to skip the objects? This would be changing `-- None --` to skip, which would better explain what the option does.

Also, I hit the following error after importing and navigating to Visualize, but not quite sure the cause:

![screenshot 2017-09-25 20 01 56](https://user-images.githubusercontent.com/40265/30840814-4ce3346e-a22e-11e7-9107-582fa020ae9b.png)

alexfrancoeur commented:
@chrisronline I think you and @tylersmalley have the right approach here. `Skip index pattern` might be a good option for now so you don't leave the context of the modal to go create an index pattern. Then you're stuck with the problem of "how do I get back to this view"?

This is definitely out of scope but I think it's useful to think through this. If I were to skip an index pattern, I would have no way of knowing which saved objects were affected after I close the modal window. It would be great if we could show a "conflict" icon in the saved object lists for any saved object that does not have a matching index pattern. We could even provide a button to "Resolve index pattern conflicts" that simply pops this modal back up. If you think this is worth pursuing, we can open a separate issue or PR.

tylersmalley commented:
To confirm, if you skip it the object will not be imported. The user would then be able to skip some objects, create an index pattern, then re-import.

alexfrancoeur commented:
@tylersmalley thanks for the clarification. Maybe rather than "Skip" we could have something a little more explicit like "Don't import" or "Skip import". I was under the impression we'd still be importing but with saved objects that result in an error.

chrisronline commented:
To clarify, the `None` option simply keeps the old, broken index pattern id instead of assigning a new one. All saved objects are still imported. 

Maybe we change the option to `Do not change`, but add some copy if this is selected that tells the user the saved object will be in a broken state until they manually update the index pattern id, or re-run the import process?

cjcenizal commented:
@chrisronline How about instead of "None" you change the option to "Fix later"?

chrisronline commented:
<img width="816" alt="screen shot 2017-09-27 at 9 02 47 am" src="https://user-images.githubusercontent.com/56682/30914743-ac33dce0-a362-11e7-836a-3a806f72c1b9.png">

@cjcenizal How does that look?

alexfrancoeur commented:
@chrisronline @cjcenizal if we're going to provide the option to `Fix later` we should have a way to do that. This goes back to my previous comment

> This is definitely out of scope but I think it's useful to think through this. If I were to skip an index pattern, I would have no way of knowing which saved objects were affected after I close the modal window. It would be great if we could show a "conflict" icon in the saved object lists for any saved object that does not have a matching index pattern. We could even provide a button to "Resolve index pattern conflicts" that simply pops this modal back up. If you think this is worth pursuing, we can open a separate issue or PR.

cjcenizal commented:
@alexfrancoeur I really like your suggestion but I'll leave it up to @chrisronline to decide whether to implement it in this PR or a later one.

@chrisronline That looks good, but TBH I think we can drop the double hyphens and the option will still be visually differentiated from the others. And I assume you're still planning on removing the empty toolbar area when there are no pagination controls?

chrisronline commented:
I agree we should provide a better UX flow for this new `Fix later` flow, but I'd like to follow up with that in another PR. 

I addressed the toolbar issue and updated the select option:
<img width="799" alt="screen shot 2017-09-27 at 4 21 23 pm" src="https://user-images.githubusercontent.com/56682/30935882-156ed182-a3a0-11e7-8b8b-611a7ea8302d.png">

Here is how it will look with pagination controls:
<img width="798" alt="screen shot 2017-09-27 at 4 22 22 pm" src="https://user-images.githubusercontent.com/56682/30935898-23e43630-a3a0-11e7-9da7-eb39576fce9b.png">

chrisronline commented:
Just an FYI, it looks like that `Fix Later` option isn't actually working so be aware of that if you pull this down and test. I'm working on fixing it up

cjcenizal commented:
UI LGTM!

tylersmalley commented:
Still getting the error when no index patterns are present:

```
TypeError: Cannot read property 'id' of undefined
    at http://localhost:5601/bundles/kibana.bundle.js?v=8467:332403:47
    at http://localhost:5601/bundles/commons.bundle.js?v=8467:3786:25
    at http://localhost:5601/bundles/commons.bundle.js?v=8467:3464:16
    at baseForOwn (http://localhost:5601/bundles/commons.bundle.js?v=8467:2437:15)
    at http://localhost:5601/bundles/commons.bundle.js?v=8467:3785:10
    at new ChangeIndexModal (http://localhost:5601/bundles/kibana.bundle.js?v=8467:332401:39)
    at http://localhost:5601/bundles/kibana.bundle.js?v=8467:128129:19
    at measureLifeCyclePerf (http://localhost:5601/bundles/kibana.bundle.js?v=8467:127910:13)
    at ReactCompositeComponentWrapper._constructComponentWithoutOwner (http://localhost:5601/bundles/kibana.bundle.js?v=8467:128128:17)
    at ReactCompositeComponentWrapper._constructComponent (http://localhost:5601/bundles/kibana.bundle.js?v=8467:128114:22)
```

Also, I believe there is confusion on the actual functionality of the "fix later". Choosing that will not result in the object being imported with a broken reference to the index pattern. We can change that functionality, but I don't feel comfortable backporting that to a patch release if we decide to backport this PR. We should consider alternative language for the option.

And if anyone is wondering, I have an export with 1,151 searches and it takes over 3 minutes to import... with only the activity indicator the on the top of the page letting you know things are happening. It's really an aside, but we should work on the story better - especially with Spaces. I imagine more folks will use export/import as a sharing hack.

https://www.dropbox.com/s/fhf9scpmpq5h61t/export-large.json?dl=0

![screenshot 2017-09-27 17 57 40](https://user-images.githubusercontent.com/40265/30944669-463a652a-a3ae-11e7-91d4-6918760ee9fb.png)

chrisronline commented:
@tylersmalley Thanks for pointing that out. I realized some locally commented out code gave me the wrong impression on how the import worked with missing index pattern ids, but I now see the behavior you described earlier. I entirely agree that we shouldn't backport a change like that and for this PR, we should preserve existing functionality, which is to _not_ import the saved object.

I'm suggesting we go back to `Skip import` as the label as well.

I'll also look into that error scenario more and take a pass on how we might able to handle importing large data sets.

chrisronline commented:
I fixed up the code so that the select option now reads `Skip import` and the code will respect that and not attempt to import it. I also addressed the issue around trying to import when there are no index patterns. Both of those should be ready to go.

For the long import scenario, I'm not sure there is much we can do right now. Maybe an ideal flow will be to always show a modal or something and that modal can show a loading indicator or some kind of progress meter after the import starts.

@tylersmalley good for another pass

chrisronline commented:
Ready for a final pass @tylersmalley

tylersmalley commented:
LGTM on green, @cjcenizal did you want to give the recent changes a look?

tylersmalley commented:
Failure looks legitimate, but not immediately sure if it's related.

https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/11366/

Jenkins, test this

chrisronline commented:
Backport PRs:

5.6: https://github.com/elastic/kibana/pull/14323
6.0: https://github.com/elastic/kibana/pull/14321
6.x: https://github.com/elastic/kibana/pull/14322
